domain/service/employees.go

```diff
@@ -0,0 +1,187 @@
+package service
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"time"
+
+	"github.com/Shelex/grpc-go-demo/domain/entities"
+	"github.com/Shelex/grpc-go-demo/proto"
+	"github.com/Shelex/grpc-go-demo/storage"
+	"github.com/Shelex/grpc-go-demo/storage/documents"
+	"github.com/google/uuid"
+	log "github.com/sirupsen/logrus"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/metadata"
+)
+
+type employeeService struct {
+	repository storage.Storage
+	documents  documents.FileStorage
+	grpc       *grpc.Server
+}
+
+func (e *employeeService) Employees(req *proto.GetAllRequest, stream proto.EmployeeService_EmployeesServer) error {
+	employees, err := e.repository.GetAll()
+	if err != nil {
+		return fmt.Errorf("failed to get employees: %w", err)
+	}
+	for _, emp := range employees {
+		if err := stream.Send(&proto.EmployeeResponse{Employee: entities.EmployeeFromStorageToProto(emp)}); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (e *employeeService) EmployeeByID(ctx context.Context, req *proto.ByIDRequest) (*proto.EmployeeResponse, error) {
+	if md, ok := metadata.FromIncomingContext(ctx); ok {
+		log.Printf("metadata received: %v\n", md)
+	}
+	employee, err := e.repository.GetEmployee(req.ID)
+	if err != nil {
+		return nil, err
+	}
+	return &proto.EmployeeResponse{
+		Employee: entities.EmployeeFromStorageToProto(employee),
+	}, nil
+}
+
+func (e *employeeService) AddEmployee(ctx context.Context, req *proto.EmployeeRequest) (*proto.EmployeeResponse, error) {
+	req.Employee.ID = uuid.New().String()
+	employee, err := e.repository.AddEmployee(entities.EmployeeFromProtoToStorage(req.Employee))
+	if err != nil {
+		return nil, fmt.Errorf("failed to save employee: %w", err)
+	}
+	log.Printf("employee %s successfully saved;\n", employee.ID)
+	return &proto.EmployeeResponse{Employee: entities.EmployeeFromStorageToProto(employee)}, nil
+}
+func (e *employeeService) AddEmployees(stream proto.EmployeeService_AddEmployeesServer) error {
+	var savedCount int
+	var errorMessage string
+	for {
+		emp, err := stream.Recv()
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			return err
+		}
+		emp.Employee.ID = uuid.New().String()
+		saved, err := e.repository.AddEmployee(entities.EmployeeFromProtoToStorage(emp.Employee))
+		if err != nil {
+			errorMessage += fmt.Sprintf("\n%s", err.Error())
+			continue
+		}
+		if err := stream.Send(&proto.EmployeeResponse{
+			Employee: entities.EmployeeFromStorageToProto(saved),
+		}); err != nil {
+			return err
+		}
+		savedCount++
+	}
+	if errorMessage != "" {
+		return errors.New(errorMessage)
+	}
+	log.Printf("successfully saved %d employees", savedCount)
+	return nil
+}
+
+func (e *employeeService) AddAttachment(stream proto.EmployeeService_AddAttachmentServer) error {
+	md, ok := metadata.FromIncomingContext(stream.Context())
+	var userID string
+	var fileName string
+	if ok {
```

Comment:
All this logic to get the userid and fileName could be in a separate function for clarity?

Reply:
Yeah, i think metadata operations should be moved to separate functions, the only reason it is not done for now - metadata not used functionally, only to check how it is working :)